joomla/joomla-cms
# [4.4] Invalid timings in debug plugin

### Summary of Changes

The time and memory displayed in debug bar (right position) incorrectly includes the usage of debug plugin itself: it's always much bigger than the sum of all timings in Profile tab.

We can't measure 100% correctly but at least should get closer: the most recent profile mark should have right border as closer to right as possible.

See that `afterRender` profile mark is not close the the right, hence we have incorrectly logged total time: it includes the time spent in debug plugin itself and we don't see the real state of page generation.


1. Debug plugin refactored to `Joomla\Event\SubscriberInterface`
2. Fixed problem with the time spent for `onAfterRespond` event is not counted.
3. Added `afterRespond` profiler mark.
4. Fixed problem with the time spent in `onAfterDisconnect` event by debug plugin is included in request time.
5. Remove unused `PlgSystemDebug::$totalQueries` variable
6. Change `PlgSystemDebug::sqlShowProfiles` to local variable because it's the only usage.
7. Custom MemoryCollector to display the peak memory usage before debug data collecting routines.
8. Display main memory usage with 3 digits.


### Testing Instructions

Apply patch, see more accurate timings in debug. Profile tab marks are now more proportionally to the 100% width.

### Actual result BEFORE applying this Pull Request

See issues in Changes above.

### Expected result AFTER applying this Pull Request

See issues in Changes above.

### Documentation Changes Required
No.
